Improvements to dynamic client registration

S-Luiten opened the pull request:
<!-- Provide a brief summary of your changes -->
Several improvements to dynamic client registration:
1. Added a delegate for handling the dynamic client registration response. This can be used to retrieve the client id (and client secret for confidential clients) allowing the MCP client to re-use the created client.
2. Added support for creating public clients.
3. Added support for authenticated dynamic client registration when the authorization server does not allow creating clients anonymously.

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->
- Exposing the dynamic client registration response is required to let the MCP client access the client id and client secret. Without these, the MCP client cannot re-use the client and must create a new one every time.
- The keycloak realm I'm using is not configured to allow anonymous client registration, so an initial access token is required.

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->
This has been tested with an MCP server and MCP client both running locally, using keycloak as authorization server.

## Breaking Changes
None

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [ ] My code follows the repository's style guidelines
- [ ] New and existing tests pass locally
- [ ] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->
The default OAuth client type has been set to confidential so it's not a breaking change, but public might make more sense as the default?

S-Luiten commented on src/ModelContextProtocol.Core/Authentication/ClientOAuthOptions.cs:
```diff
@@ -52,6 +52,19 @@ public sealed class ClientOAuthOptions
+    /// The implementation should save the client credentials securely for future use.
+    /// </para>
+    /// </remarks>
+    public DynamicClientRegistrationDelegate? DynamicClientRegistrationDelegate { get; set; }
```
Do you want the existing properties `ClientName` and `ClientUri` in this new class as well? It would be a breaking change though.

halter73 replied:
Yes. Good call. We're not too concerned about making these kinds of breaking changes this early.

S-Luiten commented on src/ModelContextProtocol.Core/Authentication/ClientOAuthProvider.cs:
```diff
@@ -69,6 +72,7 @@ public ClientOAuthProvider(
         _redirectUri = options.RedirectUri ?? throw new ArgumentException("ClientOAuthOptions.RedirectUri must configured.");
         _clientName = options.ClientName;
         _clientUri = options.ClientUri;
+        _clientType = options.ClientType ?? OAuthClientType.Confidential;
```
> I'm curious why you added the option for a public client though if you're going through the effort of dynamic client registration? Why not just use a secret if the server is willing to give you one?
> 
> If you're not doing dynamic client registration, you can already specify a client ID without a client secret if you want to act as a public client.

I'm only using DCR because the MCP client I was trying to get this working with does not support specifying the client ID. If you create a new client every time it doesn't really matter if it's public or confidential, but I also included the `DynamicClientRegistrationDelegate` so that the client can be re-used, and in that case the choice for public or confidential does matter.

halter73 replied:
I'm still not convinced that the choice of public or confidential matters unless there's some authorization server that supports DCR but doesn't support client secrets which I have not seen yet.

I can see how the `DynamicClientRegistrationDelegate` is useful, but if you've retained tokens from a previous session using that delegate and use those, you're no longer doing DCR for the subsequent sessions. At that point you're just a normal confidential client like any other time you manually configure `ClientOAuthOptions.ClientSecret`.

I think we should remove ClientType/OAuthClientType entirely, and just add InitialAccessToken and the DynamicClientRegistrationDelegate to the DynamicClientRegistrationOptions.

halter73 commented on src/ModelContextProtocol.Core/Authentication/ClientOAuthProvider.cs:
```diff
@@ -483,6 +525,11 @@ private async Task PerformDynamicClientRegistrationAsync(
+
+        if (_dynamicClientRegistrationDelegate is not null)
+        {
+            await _dynamicClientRegistrationDelegate(registrationResponse, cancellationToken).ConfigureAwait(false);
```
I'm okay just having it handle the response for now then.

halter73 commented:
> I agree that pre-registering a client makes a lot more sense than DCR and is also easier, but the problem is that not all MCP clients support this. For example vscode has built-in support for microsoft and github authentication, but only supports DCR for third party authorization servers, so there's no way for me to pass the client id of the pre-configured client.

The problem is that clients like VS Code are likely not going to support DCR with an initial access token for the same reason that they do not support using a preregistered client ID and client secret which is that it requires additional configuration. If VS Code ever adds the ability to specify the DCR initial access token via config, why wouldn't they also add support for specifying a client ID and client secret for your preregistered client?

S-Luiten commented:
@halter73 I think I've addressed all your feedback now, let me know if there's anything else you'd like to see changed.

For me this change doesn't have much priority anymore since pre-registering a client and passing its client id is a better solution than DCR. This has also been made possible in the latest vscode insiders build and I finally got it working that way now.

mbuchberger1967 commented:
@halter73 
sorry for mis-using the PR comment, but I followed this thread and find very interesting what you said about:
using a preregistered client ID and client secret
This is exactly what I try to achive: Connect my mcp client with a mcp server via a (in AZ Portal) pre-registered client ID and secret to avoid any user-/browser based auth flow.
But I cannot find any sample how to achieve this if I use the ModelContextProtocol.Client library and failed so far to find it out by myself. Do you have a hint/sample for me?
Thanks
